### PR TITLE
fix(asr): Copy Assist no longer crashes when the GPU lacks memory for the speech model (#4972). Principle XI.

### PR DESCRIPTION
**Pull request by skerker**
## Summary

Refs #4972.

Enabling Copy Assist on a GPU without room for the model kills the app with no dialog and no
log line — reproduced on Linux, below. That is the symptom #4972 reports from Windows. Whether it
is that reporter's failure is inferred from the matching symptom, not measured: the faulting
module was never supplied, and the 2 GB card in #5730 loaded the same model. Closing #4972 is left
to the maintainers and the reporter. Two defects addressed:

1. **The crash.** Vendored `whisper.cpp` ignores a failed weight-buffer allocation
   (`whisper_model_load()`: `buf = ggml_backend_alloc_ctx_tensors_from_buft(...)`, `if (buf) {…}`,
   no `else`) and goes on to upload weights into tensors that have no buffer. The first upload
   dereferences null in the backend — `SIGSEGV`, which no `catch` in `WhisperAsrBackend` can see.
   The load now fails instead; `whisper_init_with_params_no_state()` already frees the context and
   returns NULL, and `WhisperAsrBackend::load()` already latches the device and retries on CPU
   (#4767). The same unchecked pattern in the VAD loader is patched too (not called by AetherSDR).
   Recorded as local patch 4 in `third_party/whisper.cpp/AETHERSDR-PATCHES.md`; both sites read the
   same on upstream `master` as of 2026-09-16.
2. **The trigger.** `CopyAssistController::applyGpuDevices()` raised the default tier to
   `large-v3-turbo` whenever a usable GPU existed. `AsrGpuDevice` has carried free/total memory
   since #4988; nothing read it. The raise now also needs free memory ≥ weights + 300 MiB
   (`asrTierFitsVram()`, pure, header-inline). Only the automatic raise is gated: an explicit
   model choice, a device whose memory is unknown, and a tier that is already running behave as
   before. When the raise is withheld, one warning line says why.

The issue thread and #5190 attribute the death to `GGML_ABORT`. It is a null dereference: ggml
reports the failed allocation correctly (returns NULL) and whisper drops it.

**Windows release binaries do not compile `whisper.cpp`.** They link the prebuilt `whisper.lib`
from the `whisper-gpu-1.9.1` release asset (`ASR_USE_PREBUILT_WHISPER_GPU`). Commit 1 reaches
them when that pack is rebuilt from a tree containing it and its pinned SHA-256 is updated.
Commit 2 is AetherSDR code and ships in every build: a 2 GB card is no longer handed the
1.6 GB tier.

### Requirements → hunks

| From the thread | Hunk |
|---|---|
| Enable must not kill the app on a small GPU | `whisper.cpp` allocation check → existing CPU retry |
| "check available video memory before offering the large tier" (NF0T) | `asrTierFitsVram()` + `applyGpuDevices()` |
| "fall back to CPU when it will not fit" (NF0T) | already on main (#4767), reachable once the load fails cleanly |

## Constitution principle honored

Principle XI — the crash is reproduced on main `85fdf816` and shown gone on this head, same
machine, same memory pressure (no commit touches `src/asr`, `CopyAssistController` or
`third_party/whisper.cpp` between `85fdf816` and this branch's base `57b80ed4`). Principle VIII — every figure in the code comments and tests is a
measurement with its source named.

## Test plan

- [x] Local build passes (`cmake --build build`) — Linux, clean RelWithDebInfo, Build SHA `0bbdcb5a`
- [x] Behavior verified on a real radio if applicable — **n/a**, no radio surface; bench ran on the demo simulator, RX only
- [x] Existing tests pass (CI) — local: `ctest -R 'asr_|copy_assist|voice_mode'` 11/11 on this head
- [x] Reproduction steps documented if user-reported bug — below

**Test.** `asr_gpu_probe_test` (existing target) gains the gate's cases: three MEASURED fixtures
(1126 MB and 7360 MB free from this bench; 1809 of 2176 MB from the #5730 reporter's log) and
three CONSTRUCTED ones (the exact boundary, unknown memory, unknown tier size). Mutation: with the
comparison forced true the test exits 1 — `[FAIL] 1126 MB free was judged enough for the 1.6 GB
tier (#4972)`; restored, it passes. It runs on `full-suite.yml` at merge and weekly, and on the
macOS PR lane that already runs this target.
The allocation check has no socket-free seam — forcing the failure needs a GPU backend short of
memory — so it is covered by the bench below and by the refresh procedure in `AETHERSDR-PATCHES.md`.

**Proof** (Linux, RTX 5060 Laptop 8151 MiB, driver 595.84, ggml-vulkan, demo simulator; device
memory occupied by a small Vulkan ballast program, source in the bundle):

| Build | Free memory the app logged | Enable |
|---|---|---|
| main `85fdf816` | 7360 MB | large-v3-turbo loads; the process's device memory grows by 1817 MiB (whisper's own load log sums to 1818 MiB) |
| main `85fdf816` | 1766 MB | loads and transcribes |
| main `85fdf816` | 1126 MB / 1113 MB | **SIGSEGV, n=2**: `alloc_tensor_range: failed to allocate Vulkan1 buffer of size 551900160` then kernel `AsrWorker: segfault at 10 … AetherSDR[269ab2d]` → `ggml_vk_buffer_write_2d` (`ggml-vulkan.cpp:7927`) |
| this head `0bbdcb5a` | 1126 MB | tier stays **Base**; `WRN aether.gui: ASR: keeping the default model tier - "NVIDIA GeForce RTX 5060 Laptop GPU" has 1126 MB free, "large-v3-turbo" needs about 1849 MB`; Base loads on Vulkan1 and transcribes |
| this head `0bbdcb5a` | 1098 MB, large-v3-turbo chosen explicitly | same `failed to allocate Vulkan1 buffer of size 551900160`, then `whisper_model_load: failed to allocate Vulkan1 buffer for the model weights` → `WRN GPU model load failed "(null context)" - retrying on CPU` → `GPU device 1 marked unusable for the rest of this session` → model loads on CPU, then on the other usable device (Vulkan0); **process alive, 0 kernel segfault lines** |
| this head `0bbdcb5a` | 7424 MB (no ballast) | still auto-raises to large-v3-turbo on Vulkan1, no gate warning |

Help → About read `0bbdcb5a` before the arms.

**What the logs now show** (from the bundle; before, the log simply stopped):

```
# gate — the 1.6 GB tier is withheld, and the log says why
[23:35:27.011] INF aether.asr.whisper: GPU device 1 "NVIDIA GeForce RTX 5060 Laptop GPU" - VRAM free 1126 of 8151 MB - usable
[23:35:27.015] WRN aether.gui: ASR: keeping the default model tier - "NVIDIA GeForce RTX 5060 Laptop GPU" has 1126 MB free, "large-v3-turbo" needs about 1849 MB

# explicit large-v3-turbo on the same card — the load fails and recovers instead of SIGSEGV
whisper_model_load: failed to allocate Vulkan1 buffer for the model weights
[23:37:46.233] WRN aether.asr.whisper: GPU model load failed "(null context)" - retrying on CPU
[23:37:46.233] WRN aether.asr.whisper: GPU device 1 marked unusable for the rest of this session; it will not be tried again
[23:37:46.920] INF aether.asr.whisper: Loaded model "~/.local/share/AetherSDR/models/ggml-large-v3-turbo.bin" ( 4 threads )
```

The timestamped lines are AetherSDR's log file and appear in a support bundle. The `whisper_model_load:` line is whisper's own stderr: AetherSDR installs no `whisper_log_set()` / `ggml_log_set()` callback, so whisper and ggml messages reach a terminal only.

**Evidence bundle:** head `0bbdcb5a`, before = main `85fdf816` · whole app logs, terminal captures, kernel lines, both configure logs, the ballast source, README with the claim map · [aethersdr-pr5773-asr-vram-gate-2026-09-16.zip](https://github.com/user-attachments/files/32322950/aethersdr-pr5773-asr-vram-gate-2026-09-16.zip)

**What I tried to break.** Disable → Enable after the device was latched: the log shows
`GPU device 1 … not offered`, the reload goes to Vulkan0, the process stays up. The stored
`CopyAssist` document is unchanged by the fallback (the latch is session-only, as #4767 designed
it). NULL from the allocator is not always a failure
("all tensors already allocated"), so the check tests for an unbacked tensor rather than the NULL.

**Not tested.** Windows and macOS — this bench is Linux/NVIDIA only; whether WDDM pages rather
than failing the allocation is unmeasured, and the reporter's faulting module was never supplied.
The sustained-decode crash in #5730 did not reproduce on this GPU and is not addressed here. No
settings-dialog warning is added for an explicit choice that will not fit.

**Limitation, measured.** After an explicit choice fails to load, recovery moves to the next
device with the same model; it does not try a smaller model on the same GPU. On this machine
that put large-v3-turbo on an integrated GPU, which could not keep up — the panel's Queue reading
rose past 26 s. The process survives; the backlog is #5730's subject.

## Checklist

- [x] Commits are signed (`docs/COMMIT-SIGNING.md`)
- [x] No new flat-key `AppSettings` calls — **n/a**, no settings touched
- [x] Code is clean-room (Principle IV)
- [x] All meter UI uses `MeterSmoother` — **n/a**, no meter UI
- [x] Documentation updated if user-visible behavior changed — `AETHERSDR-PATCHES.md` and
      `AETHER_VENDORING.md` record the vendored change; `CHANGELOG.md` untouched
- [x] Security-sensitive changes reference a GHSA if applicable — **n/a**

— authored by agent (Claude Code) on behalf of @skerker

## Maintainer correction and validation

The VRAM gate now also requires total capacity to cover the model, headroom, and a 512 MiB desktop reserve (`17fac3be`), including when Vulkan reports free == total without its memory-budget extension. Failed model loads now use `whisper_free()` instead of raw `delete`, releasing tensor contexts and prior successful buffer allocations before CPU retry; the new VAD allocation-failure return similarly uses `whisper_vad_free()` (`2b755bd7`).

Local macOS validation of the cleanup change: existing `asr_gpu_probe_test` and `asr_whisper_smoke_test` built and passed 2/2 headlessly. No additional tests were introduced by the maintainer correction. Allocation-failure recovery was inspected in source; the GUI bridge cannot inject that allocator failure, and no GPU-memory-pressure runtime test was performed in this pass. The earlier Linux bench remains evidence for its recorded head. The gate controls automatic promotion only: explicit model choices and a Base load with insufficient memory still rely on loader failure recovery. Windows releases still require a rebuilt/re-pinned whisper prebuilt asset to include that recovery; this PR does not claim those Windows cases are fixed.

Generated with OpenAI Codex (GPT-6 Astra)

### CODE REVIEW

**Review by NF0T — APPROVED** [review]
Approving. Independently re-derived the root cause rather than reading it
off the PR body: walked `whisper_ctx_has_unallocated_tensor()`'s test
(`t->data==NULL && t->view_src==NULL && ggml_nbytes(t)>0`) against
`ggml_backend_alloc_ctx_tensors_from_buft_impl()` in the vendored
`ggml-alloc.c` — it mirrors the allocator's own "needs allocation"
condition exactly, so this isn't a naive NULL check that could misfire on
the legitimate "nothing left to allocate" NULL. Also confirmed the cited
crash site (`ggml_vk_buffer_write_2d`, `ggml-vulkan.cpp:7924`) is real and
consistent with a null-deref at a small offset, and that the CPU-retry
path this leans on (#4767) is correctly wired in `WhisperAsrBackend::load()`.
The VRAM-gate arithmetic in `asrTierFitsVram()` checks out against
`AsrModelCatalog`'s actual tier sizes, and the test fixtures are honestly
labeled MEASURED vs. CONSTRUCTED.

Two asks, not blockers on the code as written:

**1. Please open the upstream issue against `ggml-org/whisper.cpp` — this
needs to be more than a note in `AETHERSDR-PATCHES.md`.** That file
records that master still has the defect as of 2026-09-16, but that's a
point-in-time check, not a live signal. Without an upstream-filed issue,
the next vendor refresh has no way to know this needs re-verification
except by someone remembering to read the refresh checklist before
bumping `COMMIT` — and a vendor bump is exactly the moment this patch is
silently dropped if upstream's tree has moved on without it. An upstream
issue is the thing that survives that refresh regardless of who's doing
it or whether they read our docs first, and gives us a natural signal
(issue closes) for when the local patch can come out. Worth doing for
both call sites (`whisper_model_load()` and the VAD loader) since they
share the pattern.

**2. Please open the follow-up issue for rebuilding/re-pinning the
Windows `whisper-gpu` prebuilt asset.** As documented, Windows release
binaries link the prebuilt `whisper.lib` rather than compiling this file,
so commit 1 — the actual crash fix — doesn't reach Windows until that
pack is rebuilt from a tree containing this patch and its SHA-256 is
re-pinned. Both #4972's and #5730's reporters are on Windows, so this is
the platform where the fix matters most and currently doesn't land. I'd
rather this be tracked explicitly than assumed to follow automatically
once this merges.

Note on merge authority: this PR touches
`third_party/whisper.cpp/AETHERSDR-PATCHES.md` and `AETHER_VENDORING.md`,
which fall under the repo's `*.md` CODEOWNERS rule
(`@aethersdr/infrastructure`) rather than the broader source-tree rule
the rest of the diff falls under — `third_party/` isn't reclaimed back
to the source tier the way `docs/`, `resources/`, and `tests/` are. This
approval covers the source/test/vendored-source files; the two doc files
still need an infrastructure owner's sign-off before this is mergeable.

---
👨🏼‍💻 Co-authored by Claude Sonnet 5

**Review by NF0T — DISMISSED** [review]
Correcting my review above to Request Changes. The `aethersdr-agent` review
posted several hours before mine (07:07 UTC) caught a real blocker that my
own review missed, and I want to credit it specifically rather than just
change my vote silently.

**Blocker, independently re-verified, not taken on the bot's word:**
`asrTierFitsVram()`'s "free" input can silently equal "total" — meaning
the gate degrades to a total-capacity check with no real free-memory
signal at all.

Traced it in the vendored source:
`ggml_backend_vk_get_device_memory()` (`ggml-vulkan.cpp:17012-17041`)
only computes a genuine free figure when the device advertises
`VK_EXT_MEMORY_BUDGET_EXTENSION_NAME`, checked once at enumeration time
(`:7063-7069`) and cached in `device_supports_membudget[]`. When that
extension isn't present:

```cpp
if (membudget_supported && i < budgetprops.heapUsage.size()) {
    *free += budgetprops.heapBudget[i] - budgetprops.heapUsage[i];
} else {
    *free += heap.size;   // free == total
}
```

That's not a corner case chosen to be pessimistic — it's plausible on
exactly the hardware #4972 and #5730 came from. The PR's own test fixture
has a GTX 1050 reporting 2176 MiB total; on a driver without the budget
extension, that card reports 2176 MiB *free* too, which clears the 1849
MiB threshold (1549 MiB weights + 300 MiB headroom) and the auto-raise
proceeds regardless of what's actually free once Windows desktop
compositing and AetherSDR's own rendering are subtracted.

Combined with the Windows-prebuilt gap already noted in my first review:
on a Windows box with this driver characteristic, neither defense holds —
the gate passes on a fabricated reading, and the underlying null-deref
fix in `whisper.cpp` isn't even compiled into the shipped binary. On that
platform this PR can currently still reproduce #4972 as filed.

I don't think this needs a large redesign — checking total capacity
against need *in addition to* the free-memory check (so a card is refused
under either reporting mode) is a small, targeted fix, and the bot's
review already sketches it. But it should land before this merges, not
after, given it's the exact hardware class the issue reports came from.

My two asks from the first review stand independently of this — the
upstream issue filing and the Windows-prebuilt-pack follow-up are still
worth doing regardless of how the blocker above is resolved.

---
👨🏼‍💻 Co-authored by Claude Sonnet 5

File: src/asr/WhisperAsrBackend.h [hunk 1]
```diff
@@ -149,6 +149,32 @@ inline AsrTierResolution asrReconcileDefaultTier(const QString& currentTier,
...
+// report shared system memory and pass on their own numbers. Header-inline and
+// whisper-free, like asrReconcileDefaultTier above.
+inline constexpr quint64 kAsrTierVramHeadroomBytes = 300ull * 1024ull * 1024ull;
+
+inline bool asrTierFitsVram(quint64 vramFreeBytes, quint64 vramTotalBytes, qint64 tierSizeBytes)
+{
+    if (vramTotalBytes == 0 || tierSizeBytes <= 0) {
+        return true;
+    }
+    return vramFreeBytes >= static_cast<quint64>(tierSizeBytes) + kAsrTierVramHeadroomBytes;
```
Comment (jensenpat) [inline src/asr/WhisperAsrBackend.h]: [P1] Do not treat a capacity-only Vulkan reading as available memory

`asrGpuDevices()` copies `ggml_backend_dev_memory()` directly into these fields. In the vendored Vulkan implementation (`ggml_backend_vk_get_device_memory`, lines 17012–17042), devices without the memory-budget extension get `*free += heap.size`, so free equals total regardless of actual use. Consequently a reported 2176 MiB capacity passes this approximately 1849 MiB threshold even when actual free memory is below it, and `applyGpuDevices()` automatically selects Turbo. This is a demonstrated input/provenance gap, not proof that the original reporter's driver lacks the extension.

It matters especially because the Windows release selects the unchanged prebuilt whisper pack and does not compile the new loader guard. Handle capacity-only/unknown-budget readings conservatively before automatic promotion (for example, carry reporting quality and withhold promotion without a trustworthy budget, or adopt a justified capacity reserve). Preserve explicit operator choices. Add the no-budget input case to the regression fixtures when fixing the behavior.

File: third_party/whisper.cpp/src/whisper.cpp [hunk 2]
```diff
@@ -1855,6 +1869,14 @@ static bool whisper_model_load(struct whisper_model_loader * loader, whisper_con
...
             size_t size_main = ggml_backend_buffer_get_size(buf);
             WHISPER_LOG_INFO("%s: %12s total size = %8.2f MB\n", __func__, ggml_backend_buffer_name(buf), size_main / 1e6);
+        } else if (whisper_ctx_has_unallocated_tensor(ctx)) {
+            // AetherSDR local patch (#4972): upstream ignores this failure and
+            // goes on to upload weights into tensors with no buffer — a null
+            // dereference in the backend's set_tensor (SIGSEGV on a GPU that is
+            // short of memory). Fail the load instead; the caller frees the
+            // context and returns NULL.
+            WHISPER_LOG_ERROR("%s: failed to allocate %s buffer for the model weights\n", __func__, ggml_backend_buft_name(buft));
+            return false;
```
Comment (jensenpat) [inline third_party/whisper.cpp/src/whisper.cpp:1879]: [P2] Release model resources on the newly recoverable allocation failure

This return reaches `whisper_init_with_params_no_state()` lines 3759–3763, which closes the loader and calls only `delete ctx`. Neither `whisper_context` nor `whisper_model` has an owning destructor: `model.ctxs` and `model.buffers` are vectors of raw pointers. Deleting their vectors does not call `ggml_free()` or `ggml_backend_buffer_free()`; those loops exist only in `whisper_free()` (lines 3883–3896). Thus every new failed load leaks the created tensor contexts, and failure after a preceding buffer-type group succeeded also leaks that group's weight buffers while the caller retries on CPU. The allocator frees only buffers from its own failed range allocation, not preceding ctx-map groups.

Route this failure through full context cleanup (the existing `whisper_free(ctx)` supports null state), and likewise release `vctx` on the new VAD failure return. Update the patch documentation's claim that the existing delete already frees everything. This is source-traced; no runtime leak measurement was performed.

**Review by jensenpat — CHANGES_REQUESTED** [review]
Request changes — reviewed `0bbdcb5a3e9b4660224b835030fe876023eb7c02`.

Issue fit: the two changes address the measured model-load crash and the automatic tier selection discussed in #4972. All six files are in scope; no unrelated changes, new radio commands, architecture expansion, or CHANGELOG entry. The issue's latest reporter reply says the symptom stopped after an upgrade; that does not invalidate the independently described Linux reproduction, but it leaves attribution to that reporter unconfirmed.

| Scope | Review |
|---|---|
| `WhisperAsrBackend.h` | VRAM predicate and headroom; blocker on the provenance of the free-memory value |
| `CopyAssistController.cpp` | Device lookup and automatic promotion gate; explicit choices and already-active tiers preserve existing reconciliation behavior |
| `asr_gpu_probe_test.cpp` | Six arithmetic fixtures in the existing registered target; no socket owner or firmware peer introduced |
| `third_party/whisper.cpp/src/whisper.cpp` | Both allocation checks inspected against ggml allocator and caller cleanup; new recoverable path leaks allocations |
| `AETHERSDR-PATCHES.md`, `AETHER_VENDORING.md` | Local patch recorded with Windows limitation; cleanup claim needs correction with the fix |

Findings are inline: (1) total capacity masquerades as free memory on the no-budget Vulkan path; (2) failed-load cleanup does not release model allocations. Finding 1 independently confirms the existing review concern; finding 2 also contradicts the assertion that deleting the context frees its ggml resources.

Verification: inspected the complete diff, issue discussion, allocator range-failure/no-allocation paths, context ownership, CPU retry/session latch, tier reconciliation, catalog size, target registration, and Windows prebuilt selection. The allocation predicate correctly distinguishes an ordinary buffer-allocation failure from an already-allocated context. No radio-family seam changes: Flex, HL2, Icom, ANAN, RTL and Sim retain their radio command/state paths; this is shared client ASR behavior.

Hosted checks are all successful. Inspected the macOS CI log for run 35191442217: its actual checkout was synthetic merge `142cb83` of this head into `57b80ed4`; `asr_gpu_probe_test` passed 1/1 and explicitly printed the new #4972 gate assertion. That proves the existing fixtures, not Windows no-budget reporting or allocation cleanup. The Linux GPU bench in the PR remains author-reported evidence, not independently repeated here.

Local tests not run; zero configure/build calls, app launches, or delegated agents. No cold Mac build was justified once these source-level blockers were established. Current main was fetched (`e33adffa`), and the three affected production files have no intervening changes from the PR base. A local merge-tree attempt was blocked by sandbox object-write permissions; integration is not certified. GitHub still reports CHANGES_REQUESTED and unresolved threads, so no merge was attempted. Review cost: approximately 5 minutes; token/dollar cost unavailable.

Non-blocking cross-seam follow-ups: exercise controller promotion/re-entry with low-memory device records and allocator failure through CPU fallback using an injected, socket-free allocation failure. Cover already-allocated and partially allocated contexts, and verify resources are released. These extra tests are recommendations, not independent merge prerequisites. Windows prebuilt rebuild/re-pin and real Windows GPU validation remain necessary to claim that platform ships the loader fix.

**Comment by skerker** [comment]
@NF0T @jensenpat — for the vendor-refresh checklist: the unchecked weight-buffer allocation is known upstream — ggml-org/whisper.cpp#3386, with open PR ggml-org/whisper.cpp#4055 covering `whisper_model_load()` (not the VAD loader); both sites still read the same on their `master` at `5670d5c0` (2026-09-18). No new upstream issue filed.

— authored by agent (Claude Code) on behalf of @skerker

**Review by jensenpat — APPROVED** [review]
Approved at 2b755bd7ad952ddffd880f78b5e9c18e5b4ce32c after the requested corrections.

The author's 17fac3be adds the total-capacity reserve and covers the capacity-only Vulkan report. The signed cleanup correction 2b755bd7 uses whisper_free() on failed model loads and whisper_vad_free() on the new VAD allocation failure, releasing raw ggml contexts and previously allocated buffer groups before CPU retry. Patch documentation and the PR body now describe the cleanup and Windows prebuilt boundary accurately. All prior review threads are resolved; the superseded NF0T request was dismissed with the fix evidence recorded.

Validation: the existing asr_gpu_probe_test and asr_whisper_smoke_test built and passed 2/2 on macOS with QT_QPA_PLATFORM=offscreen against the exact source now committed as 2b755bd7. No new tests were added by this correction. All PR CI checks passed on this head; current-main integration is conflict-free. Source inspection verified the partial-initialization cleanup paths. The automation bridge cannot inject this allocator failure; no new GPU-pressure runtime measurement is claimed. The earlier Linux bench remains tied to its original head. Windows still needs a rebuilt/re-pinned prebuilt library to ship the vendored loader fix, as disclosed.

No remaining blockers in the reviewed scope. No follow-up tests requested.
